gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/auth/RefreshTokenAuth2Provider.java

```diff
@@ -0,0 +1,187 @@
+import static com.google.cloud.hadoop.util.HadoopCredentialConfiguration.TOKEN_SERVER_URL_SUFFIX;
+import static com.google.common.flogger.LazyArgs.lazy;
+
+import com.google.api.client.auth.oauth2.ClientParametersAuthentication;
+import com.google.api.client.auth.oauth2.RefreshTokenRequest;
+import com.google.api.client.auth.oauth2.TokenRequest;
+import com.google.api.client.auth.oauth2.TokenResponse;
+import com.google.api.client.http.GenericUrl;
+import com.google.api.client.http.HttpTransport;
+import com.google.api.client.json.JsonFactory;
+import com.google.api.client.json.jackson2.JacksonFactory;
+import com.google.cloud.hadoop.util.AccessTokenProvider;
+import com.google.cloud.hadoop.util.HttpTransportFactory;
+import com.google.cloud.hadoop.util.RedactedString;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.flogger.GoogleLogger;
+import java.io.IOException;
+import java.time.Instant;
+import java.time.ZoneId;
+import java.time.format.DateTimeFormatter;
+import java.util.Optional;
+import org.apache.hadoop.conf.Configuration;
+
+/**
+ * Retrieve an access token using the OAuth2 refresh token grant flow. See <a
+ * href="https://datatracker.ietf.org/doc/html/rfc6749#section-1.5">RFC 6749</a>.
+ */
+public class RefreshTokenAuth2Provider implements AccessTokenProvider {
+  private static final GoogleLogger logger = GoogleLogger.forEnclosingClass();
+  private static final JsonFactory JSON_FACTORY = JacksonFactory.getDefaultInstance();
+
+  private static final DateTimeFormatter dateFormat =
+      DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss.SSS").withZone(ZoneId.systemDefault());
+  private static final AccessToken EXPIRED_TOKEN = new AccessToken("", -1L);
+
+  private Configuration config;
+  private Optional<AccessToken> accessToken = Optional.empty();
+  private HttpTransport httpTransport;
+  private Optional<RedactedString> previousRefreshToken = Optional.empty();
+
+  @Override
+  public AccessToken getAccessToken() {
+    if (!accessToken.isPresent()) {
+      refresh();
+    }
+    return accessToken.get();
+  }
+
+  @Override
+  public void refresh() {
+    logger.atFine().log(
+        "Refreshing access-token based token. Our token is set to expire at '%s' and it is now '%s'",
+        lazy(
+            () ->
+                Instant.ofEpochMilli(
+                    accessToken.orElse(EXPIRED_TOKEN).getExpirationTimeMilliSeconds())),
+        lazy(() -> Instant.now()));
+
+    String tokenServerUrl =
+        TOKEN_SERVER_URL_SUFFIX.withPrefixes(CONFIG_KEY_PREFIXES).get(config, config::get);
+    RedactedString refreshToken =
+        AUTH_REFRESH_TOKEN_SUFFIX.withPrefixes(CONFIG_KEY_PREFIXES).getPassword(config);
+    String clientId =
+        AUTH_CLIENT_ID_SUFFIX.withPrefixes(CONFIG_KEY_PREFIXES).get(config, config::get);
+    RedactedString clientSecret =
+        AUTH_CLIENT_SECRET_SUFFIX.withPrefixes(CONFIG_KEY_PREFIXES).getPassword(config);
```

Comment:
To avoid extra calls, let's initialize this in constructor.

Reply:
Difficult to do so due to the requirement of the config. Would need to change the initialisation of the provider, which is not really an option for compatiblity reason.